From 4155cc64b078f4bfbcb80894ac72cfa809085411 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Thu, 28 Nov 2024 16:53:46 +0100
Subject: [PATCH] media: stm32: dcmipp: avoid vb2 buffer check at
 start_streaming

vb2 framework guarantee that we have a minimum amount of buffer
available when start_streaming is called hence there is no need
to check for buffer availability at this time.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Change-Id: I9dd44d149e9fb0ee124f9fa1eb5de3ed3053548d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/422028
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-pixelcap.c     | 17 +++++------------
 1 file changed, 5 insertions(+), 12 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
index a0c7bc70056c..99d06328a514 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
@@ -588,24 +588,17 @@ static int dcmipp_pixelcap_start_streaming(struct vb2_queue *vq,
 	reg_set(vcap, DCMIPP_PxFSCR(vcap->pipe_id), DCMIPP_PxFSCR_PIPEN);
 
 	/*
-	 * Start capture if at least one buffer has been queued,
-	 * otherwise start is deferred at next buffer queueing
+	 * vb2 framework guarantee that we have at least 'min_buffers_needed'
+	 * buffers in the list at this moment
 	 */
-	buf = list_first_entry_or_null(&vcap->buffers, typeof(*buf), list);
-	if (!buf) {
-		dev_dbg(vcap->dev, "Start streaming is deferred to next buffer queueing\n");
-		vcap->next = NULL;
-		vcap->state = DCMIPP_WAIT_FOR_BUFFER;
-		return 0;
-	}
-	vcap->next = buf;
+	vcap->next = list_first_entry(&vcap->buffers, typeof(*buf), list);
 	dev_dbg(vcap->dev, "Start with next [%d] %p phy=%pad\n",
-		buf->vb.vb2_buf.index, buf, &buf->paddr);
+		vcap->next->vb.vb2_buf.index, vcap->next, &vcap->next->paddr);
 
 	vcap->state = DCMIPP_RUNNING;
 
 	/* Start capture */
-	dcmipp_start_capture(vcap, buf);
+	dcmipp_start_capture(vcap, vcap->next);
 
 	return 0;
 
-- 
2.47.2

